### ISSUE TITLE
Final base of transcript reference sequences is not handled correctly on mapping

### ISSUE THREAD

**Issue by PeteCausey-Freeman** [biocommons/hgvs#481]
HI,

I have noticed that the final base of transcript (possibly other) reference sequences is mishandled when mapping c_to_g.

Considering transcript NM_000088.3
hgvs will correctly normalize  NM_000088.3:c.*1400_*1406= to NM_000088.3:c.*1400_*1406AACATAA=, and will also convert to an n. position e.g. NM_000088.3:n.5921_5927AACATAA=. However, mapping fails when trying to map to the genome. "HGVSInvalidIntervalError: start or end or both are beyond the bounds of transcript record."

Slightly different, but possibly related, Considering transcript NM_001145051.2
NM_001145051.2:c.659dupA should normalize to NM_001145051.2:c.660dupA. However NM_001145051.2:c.660A= is considered out of bounds. Although the metadata for this transcript implies that the final A base of the transcript has been added to complete the stop codon, the reference sequence (corroborated by SeqRepo) contains the A base at c.660 so the error "HGVSInvalidIntervalError: start or end or both are beyond the bounds of transcript record" is incorrect. Is this particular a conscious decision based on the metadata, or is it a bug?

Thanks

Pete

**Comment by reece** [comment]
It looks to me as though this bug has been fixed. (This version contains some unreleased patches, to be released shortly.)

```
In [1]: hgvs.__version__
Out[1]: '1.1.0.post2.dev21+h2c1d8d8d6a52'

In [2]: v = hp.parse_hgvs_variant("NM_000088.3:n.5921_5927AACATAA=")

In [3]: am37.n_to_g(v)
INFO:biocommons.seqrepo.fastadir.fastadir:Opening for reading: /usr/local/share/seqrepo/master/sequences/2016/0824/050154/1472014914.838945.fa.bgz
INFO:biocommons.seqrepo.fastadir.fastadir:Opening for reading: /usr/local/share/seqrepo/master/sequences/2016/0824/050925/1472015365.9150033.fa.bgz
Out[3]: SequenceVariant(ac=NC_000017.10, type=g, posedit=48261457_48261463=)

In [4]: am38.n_to_g(v)
INFO:biocommons.seqrepo.fastadir.fastadir:Opening for reading: /usr/local/share/seqrepo/master/sequences/2016/0827/234657/1472341617.01.fa.bgz
Out[4]: SequenceVariant(ac=NC_000017.11, type=g, posedit=50184096_50184102=)
```

Would you please send a complete example of the second bug. I can't replicate it.

**Comment by PeteCausey-Freeman** [comment]
Hi Reece,

Ignore the first part of the bug report. You are indeed correct that hgvs is handling the first variant correctly (I'm running 1.1.1). I must have been having a "moment". Sorry.

As for the second part of the report, this still appears to be valid. Here are the relevant bits of the code I ran

`
>>> primary_assembly = 'GRCh37'
>>> uta_current_version = 'uta_20171026'
>>> seqrepo_version = '2017-11-18'
>>> hgvs.__version__
'1.1.1'
>>> v = hp.parse_hgvs_variant("NM_001145051.2:c.659dupA")
>>> v2 = evm.c_to_n(v)
>>> terminal_bases = sf.fetch_seq(v2.ac,678,681)
>>> hn.normalize(v2)
SequenceVariant(ac=NM_001145051.2, type=n, posedit=680dupA)
>>> v3 = hp.parse_hgvs_variant("NM_001145051.2:c.660A=")
>>> v4 = evm.c_to_n(v3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/hgvs/assemblymapper.py", line 109, in c_to_n
    var_out = super(AssemblyMapper, self).c_to_n(var_c)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/hgvs/variantmapper.py", line 248, in c_to_n
    self._validator.validate(var_c)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/hgvs/validator.py", line 33, in validate
    return self._ivr.validate(var, strict) and self._evr.validate(var, strict)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/hgvs/validator.py", line 72, in validate
    (res, msg) = self._ref_is_valid(var)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/hgvs/validator.py", line 96, in _ref_is_valid
    var_x = self.vm.c_to_n(var) if var.type == "c" else var
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/hgvs/variantmapper.py", line 251, in c_to_n
    pos_n = tm.c_to_n(var_c.posedit.pos)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/hgvs/transcriptmapper.py", line 240, in c_to_n
    raise HGVSError("The given coordinate is outside the bounds of the reference sequence.")
hgvs.exceptions.HGVSError: The given coordinate is outside the bounds of the reference sequence.
`
v2 should normalize to 681dupA, but remains at 680dupA. As you can see from the seq_fetch, base n.681 is part of the reference sequence and is contained in SeqRepo. v3 c.c.660A= is a valid description, but will not map to n.681A= as it ought to.

I admit that the transcript is a little odd in that it (validly) lacks a 3 prime UTR and requires polyadanalytion  to complete the stop codon, but I can't see a reason immediately why the final base of the transcript is not being picked up by hgvs.

Cheers

Pete

**Comment by reece** [comment]
Confirmed. We're apparently not normalizing correctly at the very end of the sequence.

Minimal cases are easier to deal with. Do you agree that the following is the salient test?

```
(3.5) snafu$ hgvs-shell 
INFO:root:Starting hgvs-shell 1.1.2.dev2+n904ec3e058c7.d20180306
INFO:biocommons.seqrepo:biocommons.seqrepo 0.3.5
INFO:hgvs.dataproviders.seqfetcher:Using SeqRepo(/usr/local/share/seqrepo/master) sequence fetching
INFO:hgvs.dataproviders.uta:connected to postgresql://anonymous:anonymous@localhost/uta_dev/uta_20170117...

In [1]: v = hp.parse_hgvs_variant("NM_001145051.2:c.679_680insA")

In [2]: am37.c_to_n(v)
...
HGVSError: The given coordinate is outside the bounds of the reference sequence.

```

**Comment by PeteCausey-Freeman** [comment]
I agree that normalization seems to be missing the last base of the transcript.

I almost agree with the test, just need to amend the position of your stated variant.

I think the salient tests would be 
# c. position
v1 = hp.parse_hgvs_variant("NM_001145051.2:c.659_660insA")
am37.c_to_n(v1)
--> Out of bounds error
or
# n. position
v2 = hp.parse_hgvs_variant("NM_001145051.2:n.680_681insA")
am37.n_to_c(v2)
-- Out of bounds error.

I have tested both and also stepping the insertion 1 base 5 prime i.e. c.658_659insA or n.679_680insA. As suspected, the variants above both give out of bounds errors. The stepped back variants correctly normailze to dup A, but  positioned 1 base shy of the end of the transcript i.e. c.659dupA or n.680dupA rather than c.660dupA or n.681dupA.

**Comment by icebert** [comment]
I found this was caused by the inaccurate metadata of this transcript (NM_001145051.2) from RefSeq:

<img width="609" alt="screen shot 2018-03-16 at 17 41 44" src="https://user-images.githubusercontent.com/7262501/37514267-644a4d18-2942-11e8-94d1-842e6c5cf803.png">

The last exon should be 499 - 681 rather than 499 - 680. The final A base of the transcript was added to complete the stop codon, but the RefSeq seemed to forget to adjust the last exon metadata accordingly.

The SeqRepo uses the RefSeq annotations:

```python
>>> tx_identity_info = hdp.get_tx_identity_info('NM_001145051.2')
>>> tx_identity_info
['NM_001145051.2', 'NM_001145051.2', 'transcript', 21, 681, [138, 360, 182], 'TMEM247']

>>> tx_identity_info["lengths"]
[138, 360, 182]
```
The length of last exon should be 183, rather than 182. This makes the hgvs consider the length of this transcript to be 680, rather than 681.

https://github.com/biocommons/hgvs/blob/03cb886524bc368e04cf3be0d0d1372b2109a66d/hgvs/transcriptmapper.py#L78

```python
>>> tgt_len = sum(tx_identity_info["lengths"])
>>> tgt_len
680
```

This causes the missing of the last A base of this transcript in hgvs.
